### PR TITLE
Applies Promises to the internal requests

### PR DESCRIPTION
**Pull request by bikeath1337**
This is a POC for applying promises to the internal framework. This would be a precursor to exposing promises to the end user.

The Promise framework is home grown but fairly simple.

Ultimately, this PR needs to include the work of #253 or it's replacement.

### CODE REVIEW

File: evaporate.js [hunk 1]
```diff
@@ -1847,6 +1777,55 @@
         return code.length ? {code: code, msg: msg} : {};
     }
 
+    var Promise = function () {
```
Comment (jakubzitny) [inline evaporate.js]: There is a native [Promise](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise) support in [most of modern browsers](http://caniuse.com/#feat=promises), why not use that?

**Comment by bikeath1337** [comment]
Last I checked, they weren't supported in IE11. I would much rather use something built-in.

**Comment by bikeath1337** [comment]
Additionally, if Promise is available in a browser, we still have to have it available in tests.

We already have some difficult to test objects that are not available is window is undefined:

Promise
FileReader
localStorage

Would love to have a solution for these, too :)

**Comment by jakubzitny** [comment]
Yeah, IE11 is not supported, is it important for Evaporate to support it? And even if so, having native Promise in most of the browsers and some kind of workaround for IE11 would be better I think.

Also, Promise is available in tests. Tests run in Node and Node has Promise. FileReader is part of `file-api` dependency, which is not very good but should be able to serve the purpose.

And [localStorage](https://www.npmjs.com/package/localStorage) is available, but we can always just mock it and add it into `window` in [browser-env.js](https://github.com/TTLabs/EvaporateJS/blob/master/test/helpers/browser-env.js).

**Comment by bikeath1337** [comment]
I'm not arguing in any way against native support. I prefer it. I just don't want to cavalierly limit the audience of Evaporate to a specific deployment environment or browser subset only for the sake of principle.

IE11 is an absolute must-have for support.

**Comment by bikeath1337** [comment]
I did some research. It looks like RSVP's Promise is the perfect fit.

**Comment by jakubzitny** [comment]
`Q` and `Bluebird` are more popular but as long as it's `Promises/A+` compliant it's should be allright. Also `rsvp` is quite small, I think that's what you mean by perfect fit, right?

**Comment by bikeath1337** [comment]
Yes, it's lightweight and seems to meet the Promise contract. It's not as popular as bluebird but still had 1.5 M downloads in the past month. If we do this right, we could swap it out later if need be.

**Comment by jakubzitny** [comment]
I think this will be just allright.

**Comment by bikeath1337** [comment]
I've done more digging and thinking. This is going to be a big project that we've got to break down into manageable improvements.

**Pull request closed without merging**